lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -0,0 +1,345 @@
+					c.rustClientRef,
+					C.CGOPointer{
+						x:      C.uint16_t(mouseX),
+						y:      C.uint16_t(mouseY),
+						button: uint32(button),
+						down:   m.State == deskproto.ButtonPressed,
+					},
+				)); err != nil {
+					logrus.Warningf("Failed forwarding RDP input message: %v", err)
+					return
+				}
+			case deskproto.KeyboardButton:
+				if err := cgoError(C.write_rdp_keyboard(
+					c.rustClientRef,
+					C.CGOKey{
+						code: C.uint16_t(m.KeyCode),
+						down: m.State == deskproto.ButtonPressed,
+					},
+				)); err != nil {
+					logrus.Warningf("Failed forwarding RDP input message: %v", err)
+					return
+				}
+			default:
+				logrus.Warningf("Skipping unimplemented desktop protocol message type %T", msg)
+			}
+		}
+	}()
+}
+
+//export handle_bitmap
+func handle_bitmap(ci C.int64_t, cb C.CGOBitmap) C.CGOError {
+	return findClient(int64(ci)).handleBitmap(cb)
+}
+
+func (c *Client) handleBitmap(cb C.CGOBitmap) C.CGOError {
+	data := C.GoBytes(unsafe.Pointer(cb.data_ptr), C.int(cb.data_len))
+	// Convert BGRA to RGBA. It's likely due to Windows using uint32 values for
+	// pixels (ARGB) and encoding them as big endian. The image.RGBA type uses
+	// a byte slice with 4-byte segments representing pixels (RGBA).
+	//
+	// Also, always force Alpha value to 100% (opaque). On some Windows
+	// versions it's sent as 0% after decompression for some reason.
+	for i := 0; i < len(data); i += 4 {
+		data[i], data[i+2], data[i+3] = data[i+2], data[i], 255
+	}
+	img := image.NewNRGBA(image.Rectangle{
+		Min: image.Pt(int(cb.dest_left), int(cb.dest_top)),
+		Max: image.Pt(int(cb.dest_right)+1, int(cb.dest_bottom)+1),
+	})
+	copy(img.Pix, data)
+
+	if err := c.opts.OutputMessage(deskproto.PNGFrame{Img: img}); err != nil {
+		return C.CString(fmt.Sprintf("failed to send PNG frame %v: %v", img.Rect, err))
+	}
+	return nil
+}
+
+// Wait blocks until the client disconnects and runs the cleanup.
+func (c *Client) Wait() error {
+	c.wg.Wait()
+	C.free_rdp(c.rustClientRef)
+	return nil
+}
+
+func (c *Client) closeConn() {
+	c.closeOnce.Do(func() {
+		if err := cgoError(C.close_rdp(c.rustClientRef)); err != nil {
+			logrus.Warningf("Error closing RDP connection: %v", err)
+		}
+	})
+}
+
+func cgoError(s C.CGOError) error {
+	if s == nil {
+		return nil
+	}
+	gs := C.GoString(s)
+	C.free_rust_string(s)
+	return errors.New(gs)
+}
+
+//export free_go_string
+func free_go_string(s *C.char) {
+	C.free(unsafe.Pointer(s))
+}
+
+// Global registry of active clients. This allows Rust to reference a specific
+// client without sending actual objects around.
+var (
+	clientsMu    = &sync.RWMutex{}
+	clients      = make(map[int64]*Client)
+	clientsIndex = int64(-1)
+)
+
+func registerClient(c *Client) int64 {
+	clientsMu.Lock()
+	defer clientsMu.Unlock()
+	clientsIndex++
+	clients[clientsIndex] = c
+	return clientsIndex
+}
+
+func unregisterClient(i int64) {
+	clientsMu.Lock()
+	defer clientsMu.Unlock()
+	delete(clients, i)
```

Comment:
Does this also need to decrement `clientsIndex`?

Reply:
`clientsIndex` is like a sequential "ID" for clients, not reusable